src/julia.h

```diff
@@ -1494,7 +1494,13 @@ STATIC_INLINE void jl_eh_restore_state(jl_handler_t *eh)
         locks->len = eh->locks_len;
     }
 #endif
-    JL_SIGATOMIC_END();
+    jl_get_ptls_states()->defer_signal = eh->defer_signal;
+    // This should be the last since this can trigger a safepoint
+    // that throws a SIGINT.
+    jl_gc_state_save_and_set(eh->gc_state);
+    if (eh->defer_signal == 0) {
+        jl_sigint_safepoint();
```

Comment:
We can't really provide the grantee that an async signal will not be thrown between a `catch` and a `try` though, unless we garentee it can only be compiled and not interpreted. The GC safepoint can also throw the exception even when there's not a sigint safepoint.